File: IPurchasable.cpp
#include "IPurchasable.h"
#include <stdexcept>

double IPurchasable::applyDiscount(double discountPercent) const {
    if (discountPercent < 0 || discountPercent > 100)
        throw std::invalid_argument("discountPercent must be in [0, 100]");
    return getPrice() * (1.0 - discountPercent / 100.0);
}

File: IPurchasable.cs
namespace BookRecommender.Interfaces
{
    /// <summary>
    /// Интерфейс покупаемого товара
    /// </summary>
    public interface IPurchasable
    {
        double GetPrice();
        string ToDisplayString();
        double ApplyDiscount(double discountPercent);
    }
}